// react-transition-context/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTransitionContext = useTransitionContext;
exports.overallTransitionState = overallTransitionState;
exports.TransitionContext = TransitionContext;
exports.useTransitionStateEffect = useTransitionStateEffect;
exports.useTransitionStateEffectFilter = useTransitionStateEffectFilter;
exports.useAutofocusRef = useAutofocusRef;
exports.useLeftEffect = exports.useLeavingEffect = exports.useCameInEffect = exports.useEnteredEffect = exports.useAppearedEffect = exports.useEnteringEffect = exports.useAppearingEffect = void 0;

var React = _interopRequireWildcard(require("react"));

/**
 * 
 * @prettier
 */
var BaseTransitionContext = React.createContext('in');

function useTransitionContext() {
  return (0, React.useContext)(BaseTransitionContext);
}

var priority = ['out', 'leaving', 'appearing', 'entering'];

function overallTransitionState(parentState, childState) {
  for (var _i = 0; _i < priority.length; _i++) {
    var state = priority[_i];
    if (parentState === state || childState === state) return state;
  }

  return childState;
}

function TransitionContext(_ref) {
  var state = _ref.state,
      children = _ref.children;
  var parentState = useTransitionContext();
  var overallState = overallTransitionState(parentState, state);
  return React.createElement(BaseTransitionContext.Provider, {
    value: overallState
  }, children);
}

function outish(state) {
  return state === 'out' || state === 'leaving';
}

function useTransitionStateEffect(effect) {
  var nextState = useTransitionContext();
  var prevStateRef = (0, React.useRef)(null);
  var effectRef = (0, React.useRef)(effect);
  effectRef.current = effect;
  (0, React.useEffect)(function () {
    var prevState = prevStateRef.current;
    var effect = effectRef.current;
    prevStateRef.current = nextState;
    effect(prevState, nextState);
  }, [nextState]);
  (0, React.useEffect)(function () {
    return function () {
      var effect = effectRef.current;
      if (!outish(nextState)) effect(nextState, 'leaving');
    };
  }, []);
}

function useTransitionStateEffectFilter(filter) {
  return function (effect) {
    return useTransitionStateEffect(function (prevState, nextState) {
      if (filter(prevState, nextState)) effect(prevState, nextState);
    });
  };
}

var useAppearingEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return outish(prevState || 'out') && nextState === 'appearing';
});
exports.useAppearingEffect = useAppearingEffect;
var useEnteringEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return outish(prevState || 'out') && nextState === 'entering';
});
exports.useEnteringEffect = useEnteringEffect;
var useAppearedEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return prevState === 'appearing' && nextState === 'in';
});
exports.useAppearedEffect = useAppearedEffect;
var useEnteredEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return prevState === 'entering' && nextState === 'in';
});
exports.useEnteredEffect = useEnteredEffect;
var useCameInEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return nextState === 'in';
});
exports.useCameInEffect = useCameInEffect;
var useLeavingEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return !outish(prevState || 'out') && nextState === 'leaving';
});
exports.useLeavingEffect = useLeavingEffect;
var useLeftEffect = useTransitionStateEffectFilter(function (prevState, nextState) {
  return prevState === 'leaving' && nextState === 'out';
});
exports.useLeftEffect = useLeftEffect;

function useAutofocusRef() {
  var ref = (0, React.useRef)();
  useCameInEffect(function () {
    var el = ref.current;

    if (el) {
      el.focus();
      if (typeof el.select === 'function') el.select();
    }
  });
  return ref;
}
